[ADD][10.0] sale_manual_delivery

AaronHForgeFlow opened the pull request:
This is a proposal to supersede #404 

Basically this continuous with the work in that PR. Apart of minor bug fixing I introduced some features:

- Possibility yo deliver from sale order lines

The reason behind that is the following:

- Many companies have many long term sales. Thus, makes sense to deliver them at sale order line so we have one main screen to process all today's sales.
- In many implementations is the warehouse user the one that creates the delivery. It is very useful in those cases to have a direct access from inventory to the sale order lines that needs to create delivery for.
- We need to make this module compatible with others in this repo, that is why I included the module sale_stock_picking_blocking_sale_manual_delivery. Otherwise those two cannot be installed together. With the version at #404 that was not possible.

TODO:

- [x] Keep planning date in the wizard that creates the deliveries
- [x] Ensure different pickings are created where different planned dates are selected in the wizard

leemannd committed:
base module & sales_team

leemannd committed:
Add base for wizard

leemannd committed:
wizard for manual procurements

leemannd committed:
Update base computation

leemannd committed:
update to OCA guidelines

jcoux committed:
Add action_manual_procurement_wizard

jcoux committed:
Add action_manual_procurement_wizard in view

leemannd committed:
flake8 & OCA standards

jgrandguillaume committed:
[IMP] report manual procurement on SO from sale team

jgrandguillaume committed:
[ADD] Translation and .pot file

jgrandguillaume committed:
[TYP] Change manual proc for manual delivery

jgrandguillaume committed:
[IMP] Toggle manual,standard as a button in SO + UI

jgrandguillaume committed:
[IMP] Update translation files

jgrandguillaume committed:
[IMP] lot's of functionnal improvments

Don't add already planned or shipped lines in wizard
Ensure you have one picking per manual delivery
Forbid to add lines or ship more than ordered
Planned date is required
Improve views, naming and buttons
Better Readme
Default Carrier set from SO

jgrandguillaume committed:
[FIX] Do not add product of type 'service' in manual delivery creation

jgrandguillaume committed:
[ADD] Tests on various use cases

mdietrichc2c committed:
Regroup picking per date_planned / carrier + correct wizard action from SO

vrenaville committed:
[FIX] picking split criteria

vrenaville committed:
[FIX] call super in case of standard delivery

vrenaville committed:
[FIX] when multiple picking candidate available take the right one with date corresponding

[FIX] revert of renaming field

mdietrichc2c committed:
Add product name to manual delivery lines

mdietrichc2c committed:
Add SO line description

Commit:
add Delivery Address to wizard. by default selec SO partner address.

AaronHForgeFlow committed:
[ENH]sale_manual_delivery at line level

AaronHForgeFlow committed:
[ADD]sale_stock_picking_blocking_sale_manual_delivery

nikul-serpentcs commented on sale_manual_delivery/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+    'author': 'Camptocamp SA, Odoo Community Association (OCA)',
+    "license": "AGPL-3",
+    'version': '10.0.1.0.0',
+    'website': 'http://camptocamp.com',
```
@aheficent should be `https://github.com/OCA/sale-workflow`

rafaelbn commented:
Hi @aheficent , this is complicated.

Please @jgrandguillaume could you give you opinion? You validate https://github.com/OCA/sale-workflow/pull/404 but we can't merge because bots are red :red_circle:  You colleague @Tonow-c2c migrated to https://github.com/OCA/sale-workflow/pull/922 so you made efforts...

> BTW: CI red seems unrelated (7/9 builds are green)

@aheficent please explain the errors, "unrelated" don't say too much you can inform reviewers why travis fails (module, test, etc..).

AaronHForgeFlow commented:
@rafaelbn @jgrandguillaume  This PR will only supersede #404 if you agree I did not mean to change the original purpose of the module. The only thing that was removed is the following:

> The carrier and the date planned is handled in Odoo by other modules. IMHO this module should not change that because it would be overwriting what other modules does. For example, the delivery wizard is overwriting the date_planned, but this is covered in the module sale_order_line_date and sale_procurement_group_by_requested_date

This is not blocking I can add that feature back. 

In any case I consider it would be great to allow to create deliveries from sale order lines because in many cases is the warehouse user the one that creates the delivery when he creates the picking.

I will fix the CI even if it is not related to this module.

Regards.

rousseldenis commented:
@rafaelbn @aheficent I've restarted failed builds as I've fixed 10.0 main branch recently.

AaronHForgeFlow commented:
Thank you @rousseldenis

rousseldenis commented:
@aheficent Could you see runbot warning ? 

WARNING openerp_test odoo.models: ir.ui.view.create() includes unknown fields: manual_delivery_wizard_form

AaronHForgeFlow commented:
> @aheficent Could you see runbot warning ?
> 
> WARNING openerp_test odoo.models: ir.ui.view.create() includes unknown fields: manual_delivery_wizard_form

Yes, I think last commit fixes that. Thank you!

rousseldenis commented on sale_manual_delivery/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    'name': 'Sale Manual Delivery',
+    'category': 'Sale',
```
s/Sale/Sales Because it won't appear in good category on OCA shop

rousseldenis commented on sale_manual_delivery/models/sale_order.py:
```diff
@@ -0,0 +1,45 @@
+
+    @api.multi
+    def toggle_manual(self):
+        if isinstance(self.id, int):  # if already saved
```
I think the good usage is to use NewId

rousseldenis commented on sale_manual_delivery/models/sale_order_line.py:
```diff
@@ -0,0 +1,57 @@
+    @api.multi
+    def _action_procurement_create(self):
+        new_procs = self.env['procurement.order']
+        for line in self:
```
As _action_procurement_create() is multi, it is better to use it.

```suggestion
                normal_lines = self.filtered(lambda l: l.order_id.manual_delivery)
                new_procs  = super(SaleOrderLine,
                                   normal_lines)._action_procurement_create()
```

rousseldenis commented on sale_manual_delivery/models/sale_order_line.py:
```diff
@@ -0,0 +1,57 @@
+                        move.product_uom_qty,
+                        line.product_uom)
+            line.existing_qty = qty
+            if line.product_uom_qty - line.existing_qty > 0.0:
```
It's better to use float_compare

rousseldenis commented on sale_manual_delivery/models/stock_move.py:
```diff
@@ -0,0 +1,35 @@
+    def assign_picking(self):
+        # Redefine method to filter on date_expected (allows to group moves
+        # from the same manual delivery in one delivery)
+        if 'manual_delivery' not in self.env.context:
```
IMHO if passing parameters to context can be avoided it's better.

Why not putting field 'manual' on procurement and transmit it ? Moreover, if procurement_jit is not installed, you'll loose this behaviour.

rousseldenis commented on sale_manual_delivery/wizard/line.py:
```diff
@@ -0,0 +1,57 @@
+        related='order_line_id.name',
+        readonly=True
+    )
+    ordered_qty = fields.Float(
```
Use Decimal Precision here and for all qties fields

rousseldenis commented on sale_manual_delivery/wizard/manual_proc.py:
```diff
@@ -0,0 +1,84 @@
+    def onchange_line_ids(self):
+        lines = []
+        for line in self.sale_line_ids:
+            if (not line.existing_qty == line.product_uom_qty and
```
Use float_compare

rousseldenis commented on sale_manual_delivery/wizard/manual_proc.py:
```diff
@@ -0,0 +1,84 @@
+            new_procs = proc_order_obj
+
+            for line in wizard.line_ids:
+                if line.to_ship_qty > line.ordered_qty - line.existing_qty:
```
Use float_compare

rousseldenis commented on sale_manual_delivery/wizard/manual_proc.py:
```diff
@@ -0,0 +1,84 @@
+                        procurement_group_id.id)
+                    vals['product_qty'] = line.to_ship_qty
+                    new_proc = proc_order_obj.with_context(
+                        {'manual_delivery': True}).create(vals)
```
See above comment about context

rousseldenis reviewed (changes requested):
Code review. Please fix at least float_compares.

AaronHForgeFlow commented:
base 10.0 is broken :scream:

rafaelbn commented:
@aheficent I don't see it broken

![2019-10-07_17-12-50](https://user-images.githubusercontent.com/8736623/66324463-c7ffc280-e925-11e9-9e9b-a8b927f67227.png)

AaronHForgeFlow commented:
@rafaelbn yes, you are right, sorry for the misunderstanding

AaronHForgeFlow commented:
@rousseldenis @Nikul-Chaudhary Thank you for your patience. Comments are attended.

MiquelRForgeFlow commented on sale_stock_picking_blocking_sale_manual_delivery/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+{
+    "name": "Sale stock picking blocking and sale manual delivery",
```
```suggestion
    "name": "Sale Stock Picking Blocking Sale Manual Delivery",
```

MiquelRForgeFlow commented on sale_manual_delivery/__manifest__.py:
```diff
@@ -0,0 +1,24 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    'name': 'Sale Manual Delivery',
+    'category': 's/Sale/Sales',
```
??

```suggestion
    'category': 'Sales',
```

MiquelRForgeFlow commented on sale_stock_picking_blocking_sale_manual_delivery/__init__.py:
```diff
@@ -0,0 +1,6 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Eficent Business and IT Consulting Services S.L.
+# Copyright 2017 Serpent Consulting Services Pvt. Ltd.
```
no need copyrights in `__init__` files

MiquelRForgeFlow commented on sale_stock_picking_blocking_sale_manual_delivery/models/__init__.py:
```diff
@@ -0,0 +1,6 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Eficent Business and IT Consulting Services S.L.
+# Copyright 2017 Serpent Consulting Services Pvt. Ltd.
```
no need copyrights in `__init__` files

MiquelRForgeFlow commented on sale_manual_delivery/README.rst:
```diff
@@ -0,0 +1,63 @@
+Credits
+=======
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
```
```suggestion
* Odoo Community Association: `Icon <https://odoo-community.org/logo.png>`_.
```

AaronHForgeFlow commented:
@mreficent comments attended

MiquelRForgeFlow commented on sale_manual_delivery/models/sale_order.py:
```diff
@@ -0,0 +1,45 @@
+
+    @api.onchange('team_id')
+    def _onchange_team_id(self):
+        self.manual_delivery = self.team_id.manual_delivery \
```
because the `onchange` is `api.multi` by default:
```suggestion
        for so in self:
            if so.team_id:
                so.manual_delivery = so.team_id.manual_delivery
            else:
                so.manual_delivery = False
```

MiquelRForgeFlow commented on sale_manual_delivery/models/sale_order.py:
```diff
@@ -0,0 +1,45 @@
+
+    @api.multi
+    def toggle_manual(self):
+        if not isinstance(self.id, models.NewId):  # if already saved
```
`api.multi` again

```suggestion
        for so in self:
            if not isinstance(so.id, models.NewId):  # if already saved
            ...
```

HviorForgeFlow reviewed (approved):
Code review

AaronHForgeFlow commented:
Comments are welcome. For real.

nikul-serpentcs reviewed (approved):
Code Review LGTM :+1:

AaronHForgeFlow commented:
@leemannd @jgrandguillaume  as you are the original authors, I would appreciate a lot you opinion about this small refactoring. I'd like to continue migrating this module to all newer versions. Thank you so much.

grindtildeath commented on sale_manual_delivery/__manifest__.py:
```diff
@@ -1,20 +1,21 @@
+    'category': 'Sales',
     'author': 'Camptocamp SA, Odoo Community Association (OCA)',
+    "license": "AGPL-3",
     'version': '10.0.1.0.0',
```
Can you please update to 10.0.2.0.0 since the "first version" from #922 was used on some of our customer instance ?

grindtildeath reviewed (changes requested):
* I tried to test on runbot but keep encountering following error when creating a `sale.order.line`: "Sale order line False has unknown discounting type False "

* date_planned field was removed from the wizard, however it is really needed as it allows to set the scheduled date of the picking...

AaronHForgeFlow commented:
Hi @grindtildeath, thanks for reviewing. 

I understand, you need to keep compatibility with the current version you are using, I will put the date_planned back. BTW, I removed the dependency to delivery,  so I also removed the carrier in the wizard. Just let me know if you want also to keep that. The only think I actually need is the possibility to deliver from the sale order lines.

Regards.

AaronHForgeFlow committed:
[UPT]sale_manual_delivery, add back date_planned and carrier_id to manual procurement wizard

AaronHForgeFlow commented:
Added back the carried and the date planned to the wizard. Will check CI

grindtildeath commented:
@aheficent Thanks for adding this back. :hugs: 

I had to work on v11.0 version of this module to make sure we get split pickings when selecting different dates planned on the wizard. Would be great if we could integrate this change in your PR for v10.0 and your change in mine for v11.0

AaronHForgeFlow commented:
Sure @grindtildeath, I will take a look to the PR in 11.0 and add that feature here. Also will create a PR to your branch for the PR in v11

grindtildeath committed:
[IMP]Use a procurement group according to date planned

AaronHForgeFlow commented:
This is on WIP status now.  The feature of splitting pickings when selecting different dates planned on the wizard is WIP still.

AaronHForgeFlow commented:
This is out  of WIP status again :smiley: It can be reviewed. The warning `Sale order line False has unknown discounting type False` that appears on runbot when creating sale order lines seems totally unrelated to this module. Anyway I will try to find out what is that.

Edit: I confirm that the warning has nothing to do with this module, it comes up on runbot without this module installed: http://3394619-10-0-17db87.runbot3.odoo-community.org/web#min=1&limit=80&view_type=list&model=sale.order&menu_id=206&action=279

AaronHForgeFlow committed:
[FIX]ensure separate pickings for different sale orders

rousseldenis commented:
@AaronHForgeFlow @rafaelbn 

I am questionning about the flow of this module.

Did you look at https://github.com/OCA/sale-workflow/tree/10.0/sale_blanket_order ?

IMHO I think this is a better approach as the 'long term' sale is modelized and at each part of products confirmation, it creates sale orders that will use the standard flow.

What do you think about ?

AaronHForgeFlow commented:
Thanks for the feedback @rousseldenis .

IMO, that module can apply for long term sales. But this is not only about long term sales but also about material availability. And we don't want to create separate sales order for that. It is common that the product to be shipped depends on the date we receive the products from the supplier. Creating many pickings with many backorders is time consuming for users and they make mistakes. And sales department does not want many sales order for a single order that will never be repeated.

Many times we don't know what is available to ship until the day we are going to ship. This module is useful because we can choose what to deliver and when, and also who will be the carrier and what warehouse will ship the products (the route). Usually, that information is not known at the moment of sales confirmation, and creating another SO because of that is not ok.

Regards.

rousseldenis commented:
> Thanks for the feedback @rousseldenis .
> 
> IMO, that module can apply for long term sales. But this is not only about long term sales but also about material availability. And we don't want to create separate sales order for that. It is common that the product to be shipped depends on the date we receive the products from the supplier. Creating many pickings with many backorders is time consuming for users and they make mistakes. And sales department does not want many sales order for a single order that will never be repeated.
> 
> Many times we don't know what is available to ship until the day we are going to ship. This module is useful because we can choose what to deliver and when, and also who will be the carrier and what warehouse will ship the products (the route). Usually, that information is not known at the moment of sales confirmation, and creating another SO because of that is not ok.
> 
> Regards.

Ok, I've read detailed implementation and as usual you did great job guys! Makes sense.

rousseldenis reviewed (approved):
Code review

leemannd commented on sale_manual_delivery/wizard/manual_proc_view.xml:
```diff
@@ -0,0 +1,60 @@
+                    <group name="date">
+                        <field name='date_planned' required="1"/>
+                    </group>
+                    <group name="carrier">
```
From the original PR: missing `partner_id` . This is mandatory for us in order to be able to change the place where it's going to be shipped.

leemannd reviewed (changes requested):
Tested in V11 and there is one mandatory point for us that has been removed. The possibility to choose the delivery address from the sale_manual_delivery wizard.

Sidenote, the rest of it is a great work.

AaronHForgeFlow committed:
[IMP]add back the possibility to specify the delivery address for sale manual delivery

AaronHForgeFlow committed:
[IMP]allow to specify route when delivering manually

mmequignon committed:
[FIX] Restore res_partner field in the view with respect to last changes.

AaronHForgeFlow commented:
@leemannd  @grindtildeath I just cherry-picked https://github.com/OCA/sale-workflow/pull/935/commits/69478aff74f74f72cfa372f5ad984553efa9d9af  that includes the delivery address in the wizard. Also I added the possibility to specify the route (if the routes in sale order lines are activated in the system.).

Thanks for reviewing!

AaronHForgeFlow commented:
opps it seems CI is red, I will let you know when it's back green again

HviorForgeFlow commented on sale_manual_delivery/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+                ]
+            )
+            order.picking_ids = self.env['stock.picking'].search(
+                [('group_id', '=', proc_group.ids)]) if proc_group else []
```
```suggestion
                [('group_id', 'in', proc_group.ids)]) if proc_group else []
```

AaronHForgeFlow committed:
[FIX] domain in picking_ids computation

Co-Authored-By: Héctor Villarreal Ortega <hector.villarreal@forgeflow.com>

AaronHForgeFlow committed:
[UPT]join sale_manual_delivery to the exclude club

AaronHForgeFlow commented:
Ping @HviorForgeFlow

leemannd commented:
@AaronHForgeFlow Thanks for the work. I'm testing this in `V10` and I will give you a feedback before the end of the week.

AaronHForgeFlow commented:
Thank you @leemannd

i-vyshnevska commented on sale_manual_delivery/models/sale_order.py:
```diff
@@ -0,0 +1,65 @@
+    def action_manual_delivery_wizard(self):
+        self.ensure_one()
+        wizard = self.env['manual.delivery'].create({
+            'order_id': self.id,
```
you don't have this field in wizard

i-vyshnevska commented on sale_manual_delivery/wizard/manual_proc_view.xml:
```diff
@@ -0,0 +1,64 @@
+      <field name="model_id" ref="sale.model_sale_order_line" />
+      <field name="name">Create Manual Delivery</field>
+      <field name="key2">client_action_multi</field>
+      <field name="value" eval="'ir.actions.act_window,' + str(ref('action_wizard_manual_delivery'))" />
```
you can define two kinds of wizards one is generated from sale order and will have wider functionality to handle delivery per order like date, route, address, origin etc
the second one is initialized on sale order lines and in that case, it will have more strict functionality as it can get lines from different orders and partners  
in this way, we can satisfy the needs of all sides and we just need to define two forms of wizard and make some methods more generic to be able to customize them locally in case of need

AaronHForgeFlow replied:
I think it is the same thing, the sale order is just a bunch of sale order lines, each one is delivered individually. The only thing that is common in a sale order line is the partner. The other fields, like route, delivery address, schedule date could be different within a sale order line. 

If we restrict the selection of the lines we will be removing the possibility of changing that information. The only restriction in my opinion should be the partner, so only lines from the same partner should be selected.

i-vyshnevska commented on sale_manual_delivery/wizard/manual_proc.py:
```diff
@@ -0,0 +1,166 @@
+                    'order_line_id': line.id,
+                    'ordered_qty': line.product_uom_qty,
+                    'existing_qty': line.existing_qty,
+                    'to_ship_qty': line.product_uom_qty - line.existing_qty
```
no need to take in count qty_delivered?

AaronHForgeFlow replied:
I think existing_qty is already considered the delivered_qty. I will check.

AaronHForgeFlow replied:
It is

AaronHForgeFlow committed:
[FIX]sale_manual_delivery call to wizard. The default_get takes care

rousseldenis commented:
@AaronHForgeFlow What's the status of this ? Should'nt commit history a little bit cleaned ?

AaronHForgeFlow commented:
The last time I reviewed this it was working fine. But I am not using this module in this version so it is ok to me to close it unless someone else is interested.

leemannd commented:
We have been using it into other project and migrated it to 11.0 - 13.0 (AFAIR you did the migration to 12.0).
There are no chances we will start a new project in 10.0 using it.
I'm more in favor of merging it than only closing it.

AaronHForgeFlow commented:
Yes, I migrated it to 12.0 and I will keep using this module in newer versions. It is ok to me to merge this as it is now, but I am not working on this anymore

AaronHForgeFlow commented:
Do not delete branch please

leemannd commented:
@AaronHForgeFlow It's sad it has never been merged. If you re-open it I'm in favor of merging it (even if it's an old version).

AaronHForgeFlow commented:
@leemannd  I agree. I've been using this in production for years. However, I am not able to reopen this PR. Do I create a new one?

leemannd commented:
Yes reopen one and ping me on it. We'll try to find someone to have it merged.
